### DISCUSSION TITLE
Add @stream directive

### DISCUSSION THREAD

**Discussion by gopeter** [rmosolgo/graphql-ruby#3847] [Ideas]
Since we already have a `@defer` directive, a `@stream` one would be great too. I have a lot of use cases where I query hundreds of rows, and this could be speeded up by using a `@stream`, since infinite loading isn't always possible (for example: display a spreadsheet).

Here's some more information about `@stream`:

- https://graphql.org/blog/2020-12-08-improving-latency-with-defer-and-stream-directives/
- https://summit.graphql.com/slides/directives.pdf

Libraries like `urql` and `relay` are already supporting `@stream`. I think you had that in your docs a few years ago also 🤔 But maybe I'm mixing it up too.

**Comment by rmosolgo** [comment]
Hey, thanks for opening this discussion. For my part, I have considered (even attempted!) to build an implementation on the same code as `@defer` in GraphQL-Pro. If you'd be interested in trying it out, I'll try to get it released soon. Are you using a client with support for `@stream`? (Checking Relay's website, I just see "TODO"! https://relay.dev/docs/glossary/#stream)

**Comment by gopeter** [comment]
Whoa, that's great! Would be a pleasure to try this out, since I have a few apps with valid use cases (and a Pro subscription) 😄 I'm using `urql` [which supports it](https://formidable.com/open-source/urql/docs/comparison/).

Relay's support [is still experimental](https://github.com/facebook/relay/pull/3121), that's why is not mentioned in the docs (except [`@stream_connection`](https://relay.dev/docs/guided-tour/list-data/streaming-pagination/)).

I just wanted to add that I'm very impressed by your support. I'm using this gem for years and every issue I've opened was closed and resolved within days or even hours. I really appreciate this!

**Reply by rmosolgo** [reply]
Thanks for your kind words! 

>  within days or even hours

This one might take a tad longer 😅 . I'll follow up here when I have an update!

**Reply by gopeter** [reply]
Haha, for sure, no worries 😂

**Comment by rmosolgo** [comment]
👋  I just released GraphQL-Pro 1.21.0 which includes `GraphQL::Pro::Stream` ([doc](https://graphql-ruby.org/defer/stream.html)). It required some improvements in GraphQL-Ruby which I released in `graphql` 1.13.6. Please give it a try and let me know what you think!

**Reply by gopeter** [reply]
That was fast! 😄 But I didn't got it working. The request to my `@stream`ed list finishes after ~4 seconds (instead of 12-15), but it doesn't return anything:

![image](https://user-images.githubusercontent.com/1242960/150400377-8b55168c-04cb-4c3a-b304-efc9a23163ab.png)

![image](https://user-images.githubusercontent.com/1242960/150400397-6893c596-f5ff-4f1c-8991-7d8d82f60a60.png)

I'm not sure what you need to debug this, but I followed all the steps of your docs. I think the schema, the controller, and my query are the most important things:

```ruby
# frozen_string_literal: true

class GraphqlController < ApiController
  include ActionController::Live

  def execute
    context = { current_user:, current_tenant: ActsAsTenant.current_tenant }
    result = MyAppSchema.execute(
      params[:query],
      operation_name: params[:operationName],
      variables: ensure_hash(params[:variables]),
      context:
    )

    # Check if this is a deferred query:
    if (deferred = result.context[:defer])
      # Use built-in `stream_http_multipart`
      deferred.stream_http_multipart(response)
    else
      # Return a plain, non-deferred result
      render json: result
    end
  rescue StandardError => e
    Appsignal.set_error(e)
    raise StandardError, ([e.message] + e.backtrace).join($INPUT_RECORD_SEPARATOR)
  ensure
    response.stream.close
    Appsignal::Transaction.current.set_namespace("graphql")
    Appsignal::Transaction.current.set_action(params["operationName"] || "Unknown")
  end

  private

  # Handle form data, JSON body, or a blank value
  def ensure_hash(ambiguous_param)
    case ambiguous_param
    when String
      ambiguous_param.present? ? ensure_hash(JSON.parse(ambiguous_param)) : {}
    when Hash, ActionController::Parameters
      ambiguous_param
    when nil
      {}
    else
      raise ArgumentError, "Unexpected parameter: #{ambiguous_param}"
    end
  end
end
```

```ruby
# frozen_string_literal: true

class MyAppSchema < GraphQL::Schema
  use GraphQL::Subscriptions::ActionCableSubscriptions
  use GraphQL::Dataloader
  use GraphQL::Pro::Stream

  query Types::Query
  mutation Types::Mutation
  subscription Types::Subscription

  use GraphQL::Tracing::AppsignalTracing

  def self.resolve_type(_type, object, _ctx)
    type_class_name = if object.instance_of?(Hash)
      object["typename"].safe_constantize
    else
      object.class
    end

    type_class = "::Types::#{type_class_name}Type".safe_constantize

    raise ArgumentError, "Cannot resolve type for class #{object.class.name}" if type_class.blank?

    type_class
  end

  def self.id_from_object(object, _type_definition, _ctx)
    if object.instance_of?(Hash)
      object[:global_id]
    else
      object.to_global_id.to_s
    end
  end

  def self.object_from_id(global_id, _ctx)
    GlobalID::Locator.locate(global_id)
  end

  rescue_from(ActiveRecord::RecordNotFound) { |err, _obj, _args, _ctx, _field| raise GraphQL::ExecutionError, err }
end
```

```graphql
query GetArticleList(
  $archived: Boolean = false
  $order: [String!]
) {
  articles(
    orderBy: $order
    filter: {
      archived: $archived
    }
  ) {
    nodes @stream {
      id
      number
      description
      variation
    }
  }
}
```

**Reply by gopeter** [reply]
I've tried to run this query with [Insomnia](https://insomnia.rest/), which does not support streams, but gives some good insights. In fact we are getting the result we want (at least, I think that it should like this):

```

---
Content-Type: application/json
Content-Length: 34

{"data":{"articles":{"nodes":[]}}}
---
Content-Type: application/json
Content-Length: 77

{"path":["articles","nodes",0],"data":{"id":"gid://my-app/Article/51359"}}
---
Content-Type: application/json
Content-Length: 78

{"path":["articles","nodes",1],"data":{"id":"gid://my-app/Article/181163"}}
---
Content-Type: application/json
Content-Length: 77

{"path":["articles","nodes",2],"data":{"id":"gid://my-app/Article/85103"}}
---
Content-Type: application/json
Content-Length: 77

{"path":["articles","nodes",3],"data":{"id":"gid://my-app/Article/39706"}}
---
Content-Type: application/json
Content-Length: 78

{"path":["articles","nodes",4],"data":{"id":"gid://my-app/Article/184371"}}
---
```

![image](https://user-images.githubusercontent.com/1242960/150402094-fa0bfe4f-eac2-43b3-9a8a-df98daf9f09e.png)

So maybe it's a problem with `urql`. [They have an example in their repo](https://github.com/FormidableLabs/urql/blob/main/examples/with-defer-stream-directives/README.md), but it does not look any different from my urql configuration. I'll check again.

**Reply by gopeter** [reply]
I've compared the response from your server and the example urql server, and it's basically the same ...

```
---
Content-Type: application/json; charset=utf-8
Content-Length: 77

{"data":{"alphabet":[{"char":"A"},{"char":"B"},{"char":"C"}]},"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 58

{"data":{"char":"D"},"path":["alphabet",3],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 58

{"data":{"char":"E"},"path":["alphabet",4],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 58

{"data":{"char":"F"},"path":["alphabet",5],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 58

{"data":{"char":"G"},"path":["alphabet",6],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 58

{"data":{"char":"H"},"path":["alphabet",7],"hasNext":true}
---
Content-Type: application/json; charset=utf-8
Content-Length: 58

{"data":{"char":"I"},"path":["alphabet",8],"hasNext":true}
```

... but: their server returns at least one initial record. The `initialCount` argument is required, but even setting it to 0 return one initial record in the response:

![image](https://user-images.githubusercontent.com/1242960/150405127-d975fe91-81b8-4015-953a-8ddfbfc0e145.png)

Could be that the urql client needs some initial data (or at least not an empty array)? Maybe the `hasNext` information is also required?

**Reply by rmosolgo** [reply]
> `hasNext`

Ah, I think you're right! I wonder if that was added to the spec since I did `@defer`, I haven't seen it before. I'll update the code to include it!

As for `initialCount`, I don't see any restrictions on it in https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md, so I bet `hasNext` will do the trick...

**Reply by gopeter** [reply]
Great! Would you mind to implement `initialCount` as well? I want to try this one for my 3500 items list, and it would be good to show the first 50-100 results initially.

**Reply by rmosolgo** [reply]
I _expect_ `initialCount` to work, what happened when you tried it? (I see that [the spec](https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md#stream-arguments) describes it as optional (`Int`, not `Int!`), so I implemented it with a default value of `0`.) 

I just shipped graphql-pro 1.21.1 which includes `hasNext: true|false` -- let me know if it works any better with that version.

**Reply by gopeter** [reply]
Oh sorry, I thought that there is no `initialCount` because it was not mentioned in the docs :-)
It's working now, but it seems that my list is waiting for the full response before it renders an item. I'll check this again and come back to you. Am I able to delay the response for each row for `x` seconds?

**Reply by gopeter** [reply]
Yes, a little `console.log` debugging shows the difference. urql example of `nodes.length`:

<img width="476" alt="Bildschirmfoto 2022-01-20 um 22 00 03" src="https://user-images.githubusercontent.com/1242960/150421579-4f45ee72-7022-4a78-9b0d-9e681ce00d77.png">

And in my app:

![image](https://user-images.githubusercontent.com/1242960/150421643-85c3a802-3f98-453e-9db1-5c175a581325.png)

I've set `initialCount` to `1`.

**Comment by rmosolgo** [comment]
Ok... I spent today trying to get urql's streaming example working _exactly_ the same with a Ruby server, and I pretty much got there: https://github.com/rmosolgo/graphql-ruby-urql-stream-example There are some [caveats](https://github.com/rmosolgo/graphql-ruby-urql-stream-example#caveats), but it works as it should, IMO. 

The biggest thing is that urql seems to expect the first flush to contain only `---`, which Ruby doesn't do by default. So I wrote some manual streaming transport code here: https://github.com/rmosolgo/graphql-ruby-urql-stream-example/blob/master/server/server.rb#L101 (I'm pretty sure that clients aren't supposed to assume a one-to-one mapping between flushes and multipart batches, but I haven't found an authoritative source for that yet 🤷 .)

As for timing, I think this is because of debouncing on the client somewhere. If you want to add a pause, I'd recommend doing it inside the controller. For example, for Rails ActionController::Live, you might do something like this: 


```ruby 
if (deferred = result.context[:defer])
  # Stream an urql-friendly response over ActionController::Live
  # like https://github.com/FormidableLabs/urql/blob/main/examples/with-defer-stream-directives/server/index.js
  response.headers["Content-Type"] = "multipart/mixed; boundary=\"#{delimiter}\""
  # urql expects a leading boundary marker: 
  response.stream.write("---")
  deferred.deferrals.each do |deferral|
    patch = [
      "",
      "Content-Type: application/json; charset=utf-8",
      "",
      deferral.to_h.to_json
    ]
    if deferral.has_next?
      patch << "---"
    end 
    patch_str = patch.join("\r\n")
    response.stream.write(patch_str)
    # Sleep here to let the view render between flushes:
    sleep 0.1
  end
  
  response.stream.write("-----")
end 
```

Want to give something like that a try?

**Reply by gopeter** [reply]
Thank you so much for your efforts! I'm giving this a try over the weekend and let you know if it works 😊

**Reply by rmosolgo** [reply]
(I added argument docs to https://graphql-ruby.org/defer/stream#usage -- sorry that was missing before!)

**Reply by gopeter** [reply]
Great, thanks! I'm still struggling to implement your suggestion. Once I add this, the request takes forever and I'm am trying to find out why.

**Reply by gopeter** [reply]
Hmmm your `graphql-ruby-urql-stream-example` repo works fine, but trying to add this to my existing app does not work. I'm using the frontend/client of your stream-example repo, but connected to my API. That's my controller

```ruby
# frozen_string_literal: true

class GraphqlController < ApiController
  include ActionController::Live

  def execute
    context = { current_user:, current_tenant: ActsAsTenant.current_tenant }
    result = MyAppSchema.execute(
      params[:query],
      operation_name: params[:operationName],
      variables: ensure_hash(params[:variables]),
      context:
    )

    # Check if this is a deferred query:
    if (deferred = result.context[:defer])
      # Stream an urql-friendly response over ActionController::Live
      # like https://github.com/FormidableLabs/urql/blob/main/examples/with-defer-stream-directives/server/index.js
      response.headers["Content-Type"] = "multipart/mixed; boundary=\"-\""
      # urql expects a leading boundary marker:
      response.stream.write("---")
      deferred.deferrals.each do |deferral|
        patch = [
          "",
          "Content-Type: application/json; charset=utf-8",
          "",
          deferral.to_h.to_json
        ]
        patch << "---" if deferral.has_next?
        patch_str = patch.join("\r\n")

        sleep 0.5
        puts Time.now.to_i
        puts patch_str.inspect

        response.stream.write(patch_str)
      end

      response.stream.write("\r\n-----")

      # Use built-in `stream_http_multipart`
      # deferred.stream_http_multipart(response)
    else
      # Return a plain, non-deferred result
      render json: result
    end
  rescue StandardError => e
    Appsignal.set_error(e)
    raise StandardError, ([e.message] + e.backtrace).join($INPUT_RECORD_SEPARATOR)
  ensure
    response.stream.close
    Appsignal::Transaction.current.set_namespace("graphql")
    Appsignal::Transaction.current.set_action(params["operationName"] || "Unknown")
  end

  private

  # Handle form data, JSON body, or a blank value
  def ensure_hash(ambiguous_param)
    case ambiguous_param
    when String
      ambiguous_param.present? ? ensure_hash(JSON.parse(ambiguous_param)) : {}
    when Hash, ActionController::Parameters
      ambiguous_param
    when nil
      {}
    else
      raise ArgumentError, "Unexpected parameter: #{ambiguous_param}"
    end
  end
end
```

The client does not render anything until all records are streamed 🤔

**Reply by rmosolgo** [reply]
Ok, maybe it's a Rails vs. Sinatra thing. Let me rebuild `server.rb` using Rails, maybe that will shed more light on the situation 😖

**Reply by gopeter** [reply]
But maybe the sleep just does not work there. Running the client without, I always get multiple node lengths, which means that it renders parts of the list until new data has arrived:

![image](https://user-images.githubusercontent.com/1242960/151390802-fe6f9e52-8789-4b74-b98f-7c2f8b096f21.png)

![image](https://user-images.githubusercontent.com/1242960/151390848-172b80bc-3686-48e5-a589-324fb19f7265.png)

I think that this comes from urql's batching. So it seems to work somehow. But looking at my big query (3726 rows with 34 columns), it "waits" until ~3000 articles are loaded:

![image](https://user-images.githubusercontent.com/1242960/151391307-10d4b9cc-39c4-4251-85af-d8002e9a59f9.png)

But I think that this is related to `urql`s implementation now. `graphql-ruby` seems to stream everything correctly 👍  Many many thanks for you efforts! I'll walk to the maintainers of `urql` now, since your findings are very valuable for their work too!

**Reply by rmosolgo** [reply]
It _was_ a Rails bug 😖 . I added a Rails implementation to my example repo and I found the same behavior you described. It just _wasn't_ streaming the response. Turns out it's a Rack bug: https://github.com/rack/rack/issues/1619

Adding the [suggested workaround](https://github.com/rack/rack/issues/1619#issuecomment-848460528) made my example work: 

```ruby
response.headers['Last-Modified'] = Time.now.httpdate
```

So I updated the docs in 6c56cbc63. Does adding that header make your app work?

**Reply by kitten** [reply]
Regarding this section:

> The biggest thing is that urql seems to expect the first flush to contain only ---, which Ruby doesn't do by default.

We're basically directly based on the reference implementation for Incremental Delivery as part of the GraphQL over HTTP spec :v: https://github.com/graphql/graphql-over-http/blob/main/rfcs/IncrementalDelivery.md#content-type-multipartmixed

There the spec basically states:

> An initial boundary is sent marking the end of the preamble area.

So, I hope that clarifies why we're not looking for data before the first boundary. We could potentially try to detect whether the first boundary is "missing", however, I'm not sure whether that could be a spec-compliant addition 🤔

**Reply by gopeter** [reply]
> It _was_ a Rails bug 😖 . I added a Rails implementation to my example repo and I found the same behavior you described. It just _wasn't_ streaming the response. Turns out it's a Rack bug: [rack/rack#1619](https://github.com/rack/rack/issues/1619)
> 
> Adding the [suggested workaround](https://github.com/rack/rack/issues/1619#issuecomment-848460528) made my example work:
> 
> ```ruby
> response.headers['Last-Modified'] = Time.now.httpdate
> ```
> 
> So I updated the docs in [6c56cbc](https://github.com/rmosolgo/graphql-ruby/commit/6c56cbc63f0ebb610e025fd0ee4f50a99b32223c). Does adding that header make your app work?

Wow, yes, that did the trick! Thanks again!

**Reply by rmosolgo** [reply]
Thanks for clarifying, @kitten :+1: I went back to the example to replicate the scenario I was having trouble with a few days back and ... I couldn't! I found that I was able to greatly simplify my Ruby servers ([Sinatra](https://github.com/rmosolgo/graphql-ruby-urql-stream-example/commit/071e5b6f39248a3ea8015b435e0552727b5bb2eb), [Rails](https://github.com/rmosolgo/graphql-ruby-urql-stream-example/commit/5f801bfb4c8fde530d933943c8dc45c3ede38442)) and they both worked beautifully. 

I honestly have no idea what was going on the other day 😩. I must have gotten something else confused along the way. I removed my comment about that boundary issue from the readme, anyhow -- evidently it was my bad!
